File: Data Structures & Algorithms/house-robber/submission-4.java
class Solution {
  public int rob(int[] nums) {
        int[] cache = new int[nums.length];
        Arrays.fill(cache, -1);
        return dfs(nums, 0, cache);
    }

    public int dfs(int[] nums, int i, int[] cache) {
        if (i >= nums.length) {
            return 0;
        }
        if (cache[i] != -1) {
            return cache[i];
        }

        int skip = dfs(nums, i + 1, cache);

        int keep = nums[i] + dfs(nums, i + 2, cache);

        return cache[i] = Math.max(skip, keep);
    }
}

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-4.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:

        cache={}
        
        def dfs(i: int) -> int:
            if i>=len(nums):
                return 0
            if i in cache:
                return cache[i]
            
            

            include=1
            for j in range(i+1,len(nums)):
                if nums[j]>nums[i]:
                    include = max(include,1+dfs(j))
            cache[i] = include

            return cache[i]
        
        result=0
        for i in range(0,len(nums)):
            result =max(result,dfs(i))
        return result
